guides/plugins/plugins/plugin-fundamentals/programmically-sending-emails.md

```diff
@@ -0,0 +1,119 @@
+
+            <call method="setContainer">
+                <argument type="service" id="service_container"/>
+            </call>
+        </service>
+    </services>
+</container>
+```
+
+{% endcode %}
+
+We're injecting the `MailService` as an argument to the controller. This is the recommend way to programmatically send Emails. The service throws events and provides the ability to use templates instead
+of basic strings for the content.
+
+## Preparing & sending an Email
+
+Now we have the `MailService` available in our controller we can start preparing the Email to be send out. Please be aware that the `send()` method of the service is using a data validator which requires the following properties to be given in the `$data` array which we provide as the first argument:
+
+-   `recipients`
+-   `salesChannelId`
+-   `contentHtml`
+-   `contentPlain`
+-   `subject`
+-   `senderName`
+
+If one of these property is getting omitted we're running in a `ConstraintViolationException`. Based on these properties we can construct the necessary `$data` array for the method as the following:
+
+```markup
+public function send(SalesChannelContext $salesChannelContext): ?Response
+{
+    $salesChannelId = $salesChannelContext->getSalesChannelId();
+    $context = $salesChannelContext->getContext();
+
+    $content = 'Lorem ipsum dolor sit amet, consetetur sadipscing elitr';
+    $data = [
+        'salesChannelId' => $salesChannelId,
+        'recipients' => ['example@test.com' => 'readable name'],
+        'subject' => 'Programmatically sent mail',
+        'senderName' => 'admin@example.com',
+        'contentPlain' => $content,
+        'contentHtml' => $content
+    ];
+
+    $this->mailService->send($data, $context);
+
+    return $this->redirectToRoute('frontend.home.page');
+}
+```
+
+Please note that the controller action redirects to the home page after the Email was sent out successfully.
+
+### Customize the Email subject and sender name using twig variables
+
+The subject and the sender name will be parsed using a string template render which allows us to customize the these properties using Twig variables which we can provide as the third argument of the `send()` method.
+
+```markup
+ $templateData = [
+    'exampleVariable' => 'Example customer name'
+];
+$data = [
+    'salesChannelId' => $salesChannelId,
+    'recipients' => ['example@test.com' => 'readable name'],
+    'subject' => 'Hey {{ exampleVariable }}',
+    'senderName' => 'Test Email from sales channel {{ salesChannel.name }}',
+    'contentPlain' => $content,
+    'contentHtml' => $content
+];
+
+$this->mailService->send($data, $context, $templateData);
+```
+
+### Customize the Email content using a Twig template
+
+We can take the customization of the Email one step further when we're using the `$this->render()` method to render a Twig template and using it as the content for the Email.
+
+```markup
```

Comment:
Also here, please use php to indicate the syntax highlighting

Reply:
:+1: